mypy/checker.py

```diff
@@ -2147,8 +2147,120 @@ def check_assignment(self, lvalue: Lvalue, rvalue: Expression, infer_lvalue_type
                 rvalue_type = self.expr_checker.accept(rvalue)
                 if not inferred.is_final:
                     rvalue_type = remove_instance_last_known_values(rvalue_type)
+                    if self.options.check_need_type_annotation:
+                        self.check_forbidden_inference_types(inferred, rvalue, rvalue_type)
                 self.infer_variable_type(inferred, lvalue, rvalue_type, rvalue)
 
+    def check_forbidden_inference_types(self, var: Var, rvalue: Context,
+                                        rvalue_type: Type) -> None:
+        """Check for any list or dict literals that are inferred ambiguously and require annotation.
+        List and Dict literals with multiple object types require annotation:
+        >>> x = [1, 'a']  # Error
+        >>> x = {'a': 1, 'b': 'c'}  # Error
+        >>> x: List[object] = [1, 'a']  # OK
+        Unannotated List and Dict literals that contain an object of type `object` are OK:
+        >>> x = [1, 'a', object()]  # OK
+        Dict keys may be inferred to type `object` without annotation:
+        >>> x = ['a': 1, 2: 2]  # OK
+        """
+        if not self.scope.top_function() and self.options.allow_untyped_globals:
+            return
+
+        rvalue_type = get_proper_type(rvalue_type)
+        if isinstance(rvalue_type, Instance):
+            if isinstance(rvalue, ListExpr):
```

Comment:
I think it should be possible to reduce most of the duplicate between the three cases with an appropriate helper function.

Reply:
I tried this. The left over redundancy is due to the fact that the `Dict` case is handled a little differently from the `List` and `Set` case. Let me know if you think more could be done here.